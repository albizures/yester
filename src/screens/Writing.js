import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { View, TextInput, StyleSheet, KeyboardAvoidingView, Alert } from 'react-native'
import { NavigationActions, StackActions } from 'react-navigation'

import { Description, Title } from '../components'
import Container from '../components/Container'
import TopBar from '../components/TopBar'
import withUser, { shapeContextUser } from '../components/withUser'

import colors from '../utils/colors'
import http from '../utils/http'

class Writing extends Component {
  static propTypes = {
    navigation: PropTypes.object.isRequired,
    contextUser: PropTypes.shape(shapeContextUser).isRequired,
  }

  state = {
    title: this.props.navigation.getParam('question'),
    content: this.props.navigation.getParam('content', ''),
  }

  onSave = async () => {
    const { navigation } = this.props
    const { title, content } = this.state
    const storyId = navigation.getParam('storyId')

    try {
      const { data } = await http.put('/v1/stories/' + encodeURIComponent(storyId), {
        'title': title,
        'content': content,
      })

      navigation.dispatch(StackActions.reset({
        index: 0,
        actions: [
          NavigationActions.navigate({
            routeName: 'MainTab',
            action: NavigationActions.navigate({
              routeName: 'MyStory',
              action: NavigationActions.navigate({
                routeName: 'Home',
                params: {storyId: data.id},
              }),
            }),
          }),
        ],
      }))
    } catch (error) {
      // TODO add a custom response for validation type eg."string.min", "StoryAlreadyExists"
      console.log(error)
      console.log(error.response)
      Alert.alert('Something bad happpend, try again')
    }
  }

  onBack = () => {
    const { navigation } = this.props
    navigation.navigate('Home')
  }

  onChange = (name, value) => {
    this.setState({
      [name]: value,
    })
  }

  render () {
    const { name } = this.props.contextUser.user
    const { content, title } = this.state

    const action = (
      <Title
        keyName='writing.action'
        style={{ textDecorationLine: 'underline', color: colors.white }}
        onPress={this.onSave} />
    )
    const topBar = (
      <TopBar
        title='writing.topbar'
        onBack={this.onBack}
        action={action} />
    )
    return (
      <Container scroll topBar={topBar}>
        <KeyboardAvoidingView enabled behavior='position'>
          <View style={{paddingHorizontal: 29, paddingTop: 20}}>
            <TextInput
              multiline
              value={title}
              onChangeText={value => this.onChange('title', value)}
              style={styles.title}
              placeholder='Name your story...'
            />
            <Description keyName='writing.by' data={{author: name}} />
            <TextInput
              value={content}
              placeholder='Your story...'
              maxLength={1024}
              multiline
              onChangeText={value => this.onChange('content', value)}
              style={styles.textArea}
            />
          </View>
        </KeyboardAvoidingView>
      </Container>
    )
  }
}

export default withUser(Writing)

const styles = StyleSheet.create({
  title: {
    backgroundColor: 'transparent',
    fontSize: 25,
    fontWeight: 'bold',
    fontFamily: 'Karla-Regular',
    color: colors.mineShaft,
    textAlign: 'left',
  },
  textArea: {
    height: 400,
    fontFamily: 'Karla-Regular',
    fontSize: 16,
    paddingTop: 10,
    textAlignVertical: 'top',
    textAlign: 'justify',
    backgroundColor: 'transparent',
    marginTop: 30,
  },
})
