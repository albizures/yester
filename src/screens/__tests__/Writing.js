import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { Button } from 'react-native'

import Writing from '../Writing'

const navigation = { navigate: jest.fn() }

describe('src/screens/Writing.js', () => {
  it('should render', () => {
    const tree = renderer.create(<Writing navigation={navigation} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('when button to Reading is pressed', () => {
    it('should navigate to Reading screen', () => {
      const shallowWriting = shallow(<Writing navigation={navigation} />)
      const shallowTouchableHighlights = shallowWriting.find(Button)
      const shallowButtonToHome = shallowTouchableHighlights.at(0)

      shallowButtonToHome.prop('onPress')()

      expect(navigation.navigate).toHaveBeenCalledTimes(1)
      expect(navigation.navigate).toHaveBeenCalledWith('Reading')
    })
  })
})
