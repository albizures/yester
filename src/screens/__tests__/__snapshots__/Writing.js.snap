// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/screens/Writing.js should render 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#F6F7F9",
        "flex": 1,
      },
      Object {
        "backgroundColor": "#F6F7F9",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1E1437",
        "flexDirection": "row",
        "justifyContent": "center",
        "minHeight": 70,
        "paddingTop": 20,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
    >
      Writing
    </Text>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#F6F7F9",
        "paddingHorizontal": 29,
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          to Reading
        </Text>
      </View>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#313131",
            "fontFamily": "Karla-Regular",
            "textAlign": "left",
          },
          Object {
            "fontSize": 25,
            "fontWeight": "bold",
          },
          undefined,
        ]
      }
    >
      Title
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#313131",
            "fontFamily": "Karla-Regular",
            "textAlign": "left",
          },
          Object {
            "fontSize": 14,
          },
          undefined,
        ]
      }
    >
      By Luis Galvez
    </Text>
    <TextInput
      allowFontScaling={true}
      maxLength={1000}
      multiline={true}
      style={
        Object {
          "backgroundColor": "#fff",
          "fontFamily": "Karla-Regular",
          "fontSize": 16,
          "height": 400,
          "marginTop": 30,
          "paddingHorizontal": 15,
          "paddingTop": 15,
          "textAlign": "justify",
          "textAlignVertical": "top",
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
